pyannote/core/interval_tree.py

```diff
@@ -25,286 +25,84 @@
-                self.num += right.num
-                self.extent = (min(self.extent[0], right.extent[0]),
-                               max(self.extent[1], right.extent[1]))
-
-            self.extent = Segment(*self.extent)
-
-        def __repr__(self):
-            return '{extent: %s, num: %d}' % (repr(self.extent), self.num)
-
-    def length(self):
-        if self.root:
-            return self.root.metadata.num
-        else:
-            return 0
-
-    def kth(self, k):
-        """Get kth segment"""
-        if k < 0:
-            return _kth(self.root, self.root.num + k)
-        else:
-            return _kth(self.root, k)
-
-    def index(self, segment):
-        return _index(self.root, segment, 0)
+import collections as co
+import sortedcontainers as sc
+
+class SortedDict(sc.SortedDict):
+    length = dict.__len__
+
+class SortedSet(co.MutableSet):
+    "Sorted set intended only for segments."
+    def __init__(self, segments=()):
+        "Initialize sorted set from iterable of segments."
+        set_segments = set(segments)
+        self._set = set_segments
+        self._list = sc.SortedList(set_segments)
+        marks = (mark for segment in set_segments for mark in segment)
+        self._marks = sc.SortedList(marks)
+
+    def __contains__(self, segment):
+        return segment in self._set
+
+    def __iter__(self):
+        return iter(self._list)
+
+    def __len__(self):
+        return len(self._set)
+
+    length = __len__
+
+    def add(self, segment):
+        _set = self._set
+        if segment in _set:
+            return
+        _set.add(segment)
+        self._list.add(segment)
+        _marks = self._marks
+        _marks.add(segment.start)
+        _marks.add(segment.end)
+
+    def discard(self, segment):
+        _set = self._set
+        if segment not in _set:
+            return
+        _set.remove(segment)
+        self._list.remove(segment)
+        _marks = self._marks
+        _marks.remove(segment.start)
+        _marks.remove(segment.end)
 
     def extent(self):
-        if self.root:
-            return self.root.metadata.extent
+        if self._set:
+            _marks = self._marks
+            start = _marks[0]
+            end = _marks[-1]
+            return Segment(start=start, end=end)
         else:
             return Segment(start=np.inf, end=-np.inf)
 
-    def debug(self):
-        _debug(self.root, 0)
-
-    def overlapping(self, t):
-        """Get list of segments overlapping time t"""
-        return _overlapping(self.root, t)
-
-    def intersecting(self, segment):
-        return _intersecting(self.root, segment)
+    def kth(self, index):
+        return self._list[index]
 
     def co_iter(self, other):
-        for segment, other_segment in _co_iter(self.root, other.root):
-            yield segment, other_segment
+        for alpha in self._list:
+            temp = Segment(start=alpha.end, end=alpha.end)
+            iterable = other._list.irange(maximum=temp)
```

Comment:
Thinking out loud...

Would it be a good idea to also store a second list of segments sorted by end time (`._tsil`)?  

We could then define a second iterable 
```
pmet = Segment(start=alpha.start, end=alpha.start)
elbareti  = other._tsil.irange(maximum=pmet)
```

Depending on the position of `alpha` in both lists, we could use whichever iterable has the smallest length (assuming we can easily get the length of these iterables)